### PR TITLE
Add support for Unigram

### PR DESCRIPTION
**Pull request by codeofdusk**
### Link to issue number:
None.

### Summary of the issue:
This PR adds support for the Unigram Telegram client. Incoming messages are automatically read when the app is in the foreground, and message history can be reviewed with control+NVDA+1 through 0 as with Skype 7 and Miranda.

### Description of how this pull request fixes the issue:
This PR:

* Adds a new `NVDAObjects.behaviors.Monitor` class for reacting to changes in objects, and refactors `NVDAObjects.behaviors.LiveText` in terms of it.
* Adds a new `NVDAObjects.UIA.UWP_chat.UWPChatAppModule`, a base app module for UWP chat apps like Unigram. This will ease #10900 as well.
* Adds a new app module for Unigram based on the base class.
* Adds optimized versions of `childCount` and `getChild` for UIA objects.

### Testing performed:
Tested the app module. Verified that automatic reading and the review message scripts are functional.

Tested CMD (with UIA enabled) and verified that `LiveText` is still functional.

### Known issues with pull request:
There may be a slight (about two second) delay before NVDA reads messages.

### Change log entry:

== New features ==
- Added support for Unigram. (#10904)

== Changes for developers ==
- Added a new `NVDAObjects.behaviors.Monitor` class that polls for particular changes in an object. See `NVDAObjects.behaviors.LiveText` for an example implementation. (#10904)

### CODE REVIEW

**Comment by codeofdusk** [comment]
Cc @leonardder @michaelDCurran

File: source/NVDAObjects/UIA/__init__.py [hunk 1]
```diff
@@ -1350,6 +1350,20 @@ def _get_children(self):
 			children.append(self.correctAPIForRelation(UIA(windowHandle=windowHandle,UIAElement=e)))
 		return children
 
+	def getChild(self, index):
+		childrenCacheRequest = UIAHandler.handler.baseCacheRequest.clone()
+		childrenCacheRequest.TreeScope = UIAHandler.TreeScope_Children
+		try:
+			cachedChildren = self.UIAElement.buildUpdatedCache(childrenCacheRequest).getCachedChildren()
+		except COMError:
+			return super(UIA, self).children[index]
```
Comment (leonardder) [inline source/NVDAObjects/UIA/__init__.py]: Shouldn't this do:
```suggestion
			return super(UIA, self).getChild(index)

```

File: source/NVDAObjects/UIA/__init__.py [hunk 2]
```diff
@@ -1350,6 +1350,20 @@ def _get_children(self):
...
+	def getChild(self, index):
+		childrenCacheRequest = UIAHandler.handler.baseCacheRequest.clone()
+		childrenCacheRequest.TreeScope = UIAHandler.TreeScope_Children
+		try:
+			cachedChildren = self.UIAElement.buildUpdatedCache(childrenCacheRequest).getCachedChildren()
+		except COMError:
+			return super(UIA, self).children[index]
+		if not cachedChildren:
+			# GetCachedChildren returns null if there are no children.
+			raise IndexError
```
Comment (leonardder) [inline source/NVDAObjects/UIA/__init__.py]: This is confusing. some instances of getChild return None when out of range, the base implementation raises an index error. I think the former should be preferred, as we aren't indexing a list here.

File: source/appModules/unigram.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+			return
+		if self._chat:
+			self._chat.stopMonitoring()
+		self._chat = obj
+
+
+class MessagesListView(UIA):
+	_monitoring = False
+
+	def _getChildCount(self):
```
Comment (leonardder) [inline source/appModules/unigram.py]: Please override _get_childCount on NVDAObjects.UIA.UIA instead.

File: source/appModules/unigram.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+		childrenCacheRequest = UIAHandler.handler.baseCacheRequest.clone()
+		childrenCacheRequest.TreeScope = UIAHandler.TreeScope_Children
+		try:
+			cachedChildren = self.UIAElement.buildUpdatedCache(childrenCacheRequest).getCachedChildren()
+		except COMError:
+			return len(super().children)
```
Comment (leonardder) [inline source/appModules/unigram.py]:
```suggestion
			return super().childCount

```

File: source/appModules/unigram.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+		if not cachedChildren:
+			# GetCachedChildren returns null if there are no children.
+			raise IndexError
```
Comment (leonardder) [inline source/appModules/unigram.py]:
```suggestion
			return 0

```

File: source/appModules/unigram.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+		return cachedChildren.length
+
+	def _monitor(self):
```
Comment (leonardder) [inline source/appModules/unigram.py]: We really need a very simple monitor abstract class in NVDAObjects.behaviors. 

in https://github.com/nvaccess/nvda/pull/10058, I created a TextMonitor class, but even that's not abstract enough. Could you please implement something based on the TextMonitor class in #10058, but at least in such a way that every monitor can be started and stopped more or less the same and uses the same threading approach?

File: source/appModules/unigram.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+			lastMessage = None
+		while self._monitoring:
+			try:
+				message = self.getChild(self._getChildCount() - 1).name
+			except IndexError:
+				message = None
+			if message != lastMessage:
+				ui.message(message)
+				lastMessage = message
+			time.sleep(2)
```
Comment (leonardder) [inline source/appModules/unigram.py]: Please use a polling interval constant for this.

File: source/appModules/unigram.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+		self._monitoring = True
+		threading.Thread(target=self._monitor).start()
+
+	def stopMonitoring(self):
+		self._monitoring = False
+
+
+class MessageTextBox(UIA):
+	def initOverlayClass(self):
+		if not self.appModule.chat:
```
Comment (leonardder) [inline source/appModules/unigram.py]: It looks like the chat monitor can be instantiated in multiple ways. Could this be generalized somehow?

**Review by LeonarddeR — CHANGES_REQUESTED** [review]

File: source/appModules/unigram.py (hunk 8 above)
Comment (codeofdusk) [reply source/appModules/unigram.py]: I don't think so, as it depends where focus lands first (either the edit field or list view). If I just started monitoring when the list view gains focus, the user has to focus it before messages are announced, which isn't ideal.

**Comment by codeofdusk** [comment]
Actually, I think Unigram may support some UIA events (`ValueChange` on the scrollbar in the message window). However, there is definitely no such event in Dell Mobile Connect. I'll open a PR for that app and come back to this one later.

**Pull request closed without merging**